### DISCUSSION TITLE
Query app badges from Dock

### DISCUSSION THREAD

**Discussion by mcrescas** [FelixKratz/SketchyBar#317] [Ideas]
Hi, 

First of all, amazing library!! I have been using it for a while and is really amazing.

I have a proposal for a functionality that I would like to have: **notifications within sketchybar**.

There is a library called BTT to configure the touchbar to query the status of an app's badge (what the Dock shows). This is very useful because you can standardize completely to know if there are notifications in any of the apps (without having to search or rely on their apis).
They expose it through apple script, but I don't know if it is possible from there or you need to access it from the macOS API.

Would it be possible to have something similar? Right now it's the only thing I'm missing. Thanks you very much!

**Comment by mcrescas** [comment]
Adding more information: Pock also has this mechanism implemented.
I have been looking into the code, and seems to me that it is implemented in this folder https://github.com/pock/dock-widget/tree/master/Dock/Sources/Helpers/PockDockHelper.

**Reply by FelixKratz** [reply]
Just to be clear:
Do you want a way to query dock badges, or do you want to have an event that delivers new notifications? I think dock badges and notifications do not need to be coupled in any way. I could set the Dock badge of an app without sending a notification to the user notification center.

I think getting the badges of an app is very easy, and does not require accessibility access (unlike pocks code), when using the appropriate launch services apis. There might very well be a built-in commandline tool to get these informations as well. (edit: `lsappinfo -all list` will output, among a bunch of other things, the badge count in its `"StatusLabel"` property)

Implementing a badge event (that is not polling) is harder to achieve, as I am not even sure that an api for that exists. It could be done with an AX key value observer as a last resort, but that would require accessibility access.

**Reply by wojciech-kulik** [reply]
Great tip! However, I used this tool and I don't see any "StatusLabel" for Messages app, even though I've got a badge on it with "1".

```
> lsappinfo -all info -only StatusLabel Messages                                                                                                                                                                                          
"StatusLabel"=[ NULL ] 
```

**Reply by wojciech-kulik** [reply]
For some reason, it doesn't work with Messages and WhatsApp. However, it works with Slack and I found an alternative method for Messages:

```
sqlite3 ~/Library/Messages/chat.db "SELECT COUNT(guid) FROM message WHERE NOT(is_read) AND NOT(is_from_me) AND text !=''"
```

**Comment by mcrescas** [comment]
Thank you very much for your answer.

I have been playing around with `lsappinfo` and it totally fulfills what I wanted (I already have a small prototype working). 
I think reading the status of the badge is a good and quick way to detect pending notifications (if the application changes its status, of course). 
I'm still using macOS notification popups, so I don't need a specific behavior for that.

**Reply by Pe8er** [reply]
@mcrescas can you please share what you came up with? Thanks!

**Reply by mcrescas** [reply]
Hi @Pe8er ,

What I did was basically the following:

```bash
lsappinfo -all info -only StatusLabel $app_name | sed -nr 's/\"StatusLabel\"=\{ \"label\"=\"(.+)\" \}$/\1/p' 
```

This snippet of code does a pooling of the status of the badge of the corresponding App, extracting only the text (it could be a dot or a number). Then you can use the information as you want inside the bar 😄 

Hope it helps you!

**Reply by Pe8er** [reply]
Thanks a ton, made it work! 😀

<img width="501" alt="image" src="https://github.com/FelixKratz/SketchyBar/assets/770120/b290c49e-f5e7-4325-8310-96ae86a5efd1">

**Comment by zobi** [comment]
Hello, i use `lsappinfo` to count unread mail and it work's well, but to count unread sms it didn't work. I'm on MacOs Sequoia. Have you found a solution please ?
Thanks a lot

**Reply by wojciech-kulik** [reply]
https://github.com/FelixKratz/SketchyBar/discussions/317#discussioncomment-10167548

**Reply by zobi** [reply]
> [#317 (reply in thread)](https://github.com/FelixKratz/SketchyBar/discussions/317#discussioncomment-10167548)

I've try this solution but it didn't work for me : "Error : unable to open database chat.db : authorization denied"

**Reply by wojciech-kulik** [reply]
maybe with `sudo`?

**Reply by zobi** [reply]
> maybe with `sudo`?

i already try :'( 
`❯ sudo sqlite3 ~/Library/Messages/chat.db "SELECT COUNT(guid) FROM message WHERE NOT(is_read) AND NOT(is_from_me) AND text !=''";
Password:
Error: unable to open database "/Users/zobi/Library/Messages/chat.db": authorization denied`

same problem with or without sudo

**Reply by wojciech-kulik** [reply]
Weird, it works on my side. I'm also using Sequoia.

**Reply by zobi** [reply]
> Weird, it works on my side. I'm also using Sequoia.

Yes i don't understand, i try to reboot, leave apps, run apps nohting work.
If i try to ls /Users/zobi/Library/Messages/ it failed with "operation not permitted" same with sudo ...
in any case thank you for answering me ;)

**Reply by wojciech-kulik** [reply]
I guess probably a small hack to make a copy first would work, but it would be pretty slow. You can also check permissions on this file and modify them using `chmod`.

**Reply by zobi** [reply]
I found solution !
Select the main Apple menu > System Settings > Privacy & Security. Locate and select Full Disk Access from the list. Find Terminal in the list and toggle the button to the on position. Enter your admin password to make the change.

Thanks a lot for your help ;)

**Comment by 0ui** [comment]
I also had problems querying Messages and WhatsApp with `lsappinfo` only returning `NULL`, but I got this applescript working on MacOS Ventura. It does require Accessibility permission for sketchybar. Modify the `appList` to include your apps:

```applescript
set appList to {"Messages", "WhatsApp"}
set totalUnread to 0
  
tell application "System Events"
	tell process "Dock"
		try
			repeat with appName in appList
				set unreadCount to value of attribute "AXStatusLabel" of UI element appName of list 1
				if unreadCount exists then
					set totalUnread to totalUnread + unreadCount
				end if
			end repeat
			return totalUnread
		on error errMsg
			return 0
		end try
	end tell
end tell
```

**Comment by Sinjhin** [comment]
It's not as fast as `lsappinfo` but I just did this for myself with osascript and I think it should work across the board for all apps in the dock. [Repo Here](https://github.com/Sinjhin/SketchyMenu) and [post in the plugin discussion](https://github.com/FelixKratz/SketchyBar/discussions/12?sort=new#discussioncomment-14467279).

> without having to search or rely on their apis

@mcrescas it uses pure, vanilla `osascript`

My version looks like this: 
<img width="355" height="348" alt="image" src="https://github.com/user-attachments/assets/3bc687d7-16ad-4a51-843f-d1b8a6d4f678" />
